crates/typst-library/src/model/bibliography.rs

```diff
@@ -150,19 +165,108 @@ pub struct BibliographyElem {
 }
 
 impl BibliographyElem {
-    /// Find the document's bibliography.
-    pub fn find(introspector: Tracked<Introspector>) -> StrResult<Packed<Self>> {
+    /// Find the document's bibliographies.
+    pub fn find(introspector: Tracked<Introspector>) -> StrResult<Vec<Packed<Self>>> {
         let query = introspector.query(&Self::ELEM.select());
-        let mut iter = query.iter();
-        let Some(elem) = iter.next() else {
+        if query.len() == 0 {
             bail!("the document does not contain a bibliography");
         };
 
-        if iter.next().is_some() {
-            bail!("multiple bibliographies are not yet supported");
+        Ok(query
+            .into_iter()
+            .map(|elem| elem.to_packed::<Self>().unwrap().clone())
+            .collect())
+    }
+
+    #[comemo::memoize]
+    pub fn assign_citations(
+        introspector: Tracked<Introspector>,
+    ) -> FxHashMap<Span, Span> {
+        let bibliographies: Vec<Packed<Self>> = introspector
+            .query(&Self::ELEM.select())
+            .into_iter()
+            .map(|elem| elem.to_packed::<Self>().unwrap().clone())
+            .collect();
+        let citations = introspector.query(&CiteElem::ELEM.select());
+
+        let mut citation_map: FxHashMap<Span, Span> = FxHashMap::default();
+
+        // First citations from bibliographies with selectors
+        for bibliography in &bibliographies {
+            if let Smart::Custom(bibliography_selector) =
+                &bibliography.target.get_ref(StyleChain::default())
+            {
+                let bibliography_span = bibliography.span();
+                let bibliography_citations = introspector.query(&bibliography_selector);
+                for citation in bibliography_citations {
+                    citation_map.entry(citation.span()).or_insert(bibliography_span);
+                }
+            }
         }
 
-        Ok(elem.to_packed::<Self>().unwrap().clone())
+        // Find the bibliography for the remaining citations. Priority order:
+        // 1. First following auto bibliography containing the label
+        // 2. First preceding auto bibliography containing the label
+        for citation in citations {
+            if !citation_map.contains_key(&citation.span()) {
+                let citation_key = citation.to_packed::<CiteElem>().unwrap().key;
+                let citation_location = citation.location().unwrap();
+                if let Some(next_bib) = introspector
+                    .query(&Self::ELEM.select().after(citation_location.into(), false))
```

Comment:
We can rely on the introspector here to do these queries, but alternatively, we could also query for `selector.or(cite, bibliography)` at the top, to get both fused, in the correct order. I am not sure which is easier to work with, but the latter would be more efficient.